mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/task/SolveMailboxInconsistenciesService.java

```diff
@@ -0,0 +1,363 @@
+     */
+    private static class OrphanMailboxPathDAOEntry implements Inconsistency {
+        private final CassandraIdAndPath pathRegistration;
+
+        private OrphanMailboxPathDAOEntry(CassandraIdAndPath pathRegistration) {
+            this.pathRegistration = pathRegistration;
+        }
+
+        @Override
+        public Mono<Result> fix(Context context, CassandraMailboxDAO mailboxDAO, CassandraMailboxPathV2DAO pathV2DAO) {
+            return pathV2DAO.delete(pathRegistration.getMailboxPath())
+                .doOnSuccess(any -> {
+                    LOGGER.info("Fixing inconsistency for orphan mailboxPath {} - {}",
+                        pathRegistration.getCassandraId().serialize(),
+                        pathRegistration.getMailboxPath().asString());
+                    context.fixedInconsistencies.incrementAndGet();
+                })
+                .map(any -> Result.COMPLETED)
+                .switchIfEmpty(Mono.just(Result.COMPLETED))
+                .onErrorResume(e -> {
+                    LOGGER.error("Fail fixing inconsistency for orphan mailboxPath {} - {}",
+                        pathRegistration.getCassandraId().serialize(),
+                        pathRegistration.getMailboxPath().asString(),
+                        e);
+                    context.errors.incrementAndGet();
+                    return Mono.just(Result.PARTIAL);
+                });
+        }
+    }
+
+    /**
+     * The Mailbox is referenced in MailboxDAO but the corresponding
+     * reference is pointing to another mailbox in MailboxPathDao.
+     *
+     * This error can not be recovered as some data-loss might be involved. It is preferable to
+     * ask the admin to review then merge the two mailbowes together using {@link MailboxMergingTask}.
+     */
+    private static class ConflictingEntryInconsistency implements Inconsistency {
+        private final ConflictingEntry conflictingEntry;
+
+        private ConflictingEntryInconsistency(Mailbox mailbox, CassandraIdAndPath pathRegistration) {
+            boolean samePath = mailbox.generateAssociatedPath().equals(pathRegistration.getMailboxPath());
+            boolean sameId = mailbox.getMailboxId().equals(pathRegistration.getCassandraId());
+
+            Preconditions.checkState(samePath != sameId);
+
+            this.conflictingEntry = new ConflictingEntry(
+                mailbox.getMailboxId().serialize(),
+                mailbox.generateAssociatedPath().asString(),
+                pathRegistration.getMailboxPath().asString(),
+                pathRegistration.getCassandraId().serialize());
+        }
+
+        @Override
+        public Mono<Result> fix(Context context, CassandraMailboxDAO mailboxDAO, CassandraMailboxPathV2DAO pathV2DAO) {
+            LOGGER.error("MailboxDAO contains mailbox {} {} which conflict with corresponding registration {} {}. " +
+                "We recommend merging these mailboxes together to prevent mail data loss.",
```

Comment:
is there an API for that?

Reply:
The one for Ghost mailboxes (cassandra specific) do just that, how lucky we are!